src/baseline.rs

```diff
@@ -7,22 +8,203 @@ use crate::dump::RustDocCommand;
+    let placeholder_manifest_path = placeholder_build_dir.join("Cargo.toml");
+
+    // Possibly fixes https://github.com/libp2p/rust-libp2p/pull/2647#issuecomment-1280221217
+    let _: std::io::Result<()> = std::fs::remove_file(placeholder_build_dir.join("Cargo.lock"));
+
+    std::fs::write(
+        &placeholder_manifest_path,
+        toml::to_string(&placeholder_manifest)?,
+    )
+    .with_context(|| "failed to write placeholder Manifest")?;
+    std::fs::write(placeholder_build_dir.join("lib.rs"), "")
+        .with_context(|| "failed to create empty lib.rs")?;
+    Ok(placeholder_manifest_path)
+}
+
+fn generate_rustdoc(
+    config: &mut GlobalConfig,
+    rustdoc: &RustDocCommand,
+    target_root: PathBuf,
+    crate_source: CrateSource,
+) -> anyhow::Result<PathBuf> {
+    let (name, version) = match crate_source {
+        CrateSource::Registry { crate_ } => (crate_.name().to_string(), crate_.version()),
+        CrateSource::ManifestPath { .. } => unimplemented!(),
+    };
+    let (build_dir, cache_dir) = match crate_source {
+        CrateSource::Registry { .. } => {
+            let crate_identifier = format!("registry-{}-{}", slugify(&name), slugify(version));
+            let cache_dir = target_root.join("cache").join(&crate_identifier);
+
+            // We assume that the generated rustdoc is untouched.
+            // Users should run cargo-clean if they experience any anomalies.
+            if cache_dir.exists() {
+                return Ok(cache_dir.join("rustdoc.json"));
+            }
+
+            let build_dir = target_root.join(crate_identifier);
+            (build_dir, cache_dir)
+        }
+        CrateSource::ManifestPath { .. } => {
+            unimplemented!()
+        }
+    };
+
+    let placeholder_manifest = create_placeholder_rustdoc_manifest(&crate_source)
+        .with_context(|| "failed to create placeholder manifest")?;
+    let placeholder_manifest_path =
+        save_placeholder_rustdoc_manifest(build_dir.as_path(), placeholder_manifest)
+            .with_context(|| "failed to save placeholder rustdoc manifest")?;
+
+    config.shell_status("Parsing", format_args!("{name} v{version} (baseline)"))?;
+    // TODO: replace (baseline) with something else
+
+    let rustdoc_path = rustdoc.dump(
+        placeholder_manifest_path.as_path(),
+        Some(&format!("{name}@{version}")),
+        false,
+    )?;
+
+    match crate_source {
+        CrateSource::Registry { .. } => {
+            // Clean up after ourselves.
+            std::fs::create_dir_all(&cache_dir).with_context(|| "failed to create cache dir")?;
+            let cache_rustdoc_path = cache_dir.join("rustdoc.json");
+            std::fs::copy(rustdoc_path, &cache_rustdoc_path)
+                .with_context(|| "failed to copy cache dir")?;
+            std::fs::remove_dir_all(build_dir).with_context(|| "failed to remove build dir")?;
```

Comment:
Do all of the `with_context` make sense? I'm not yet proficient in using Rust errors.

Reply:
Yeah, they are fine. I just un-capitalized `manifest` throughout since it isn't a name.